internal/user/user_profile_hash_v1.go

```diff
@@ -6,54 +6,56 @@ import (
 	"io"
 
 	"github.com/pkg/errors"
-	"golang.org/x/crypto/scrypt"
+
+	"github.com/kopia/kopia/internal/crypto"
 )
 
 // parameters for v1 hashing.
 const (
 	hashVersion1 = 1
 
-	v1ScryptN    = 65536
-	v1ScryptR    = 8
-	v1ScryptP    = 1
 	v1SaltLength = 32
-	v1KeyLength  = 32
 )
 
 //nolint:gochecknoglobals
-var dummyV1HashThatNeverMatchesAnyPassword = make([]byte, v1KeyLength+v1SaltLength)
+var dummyV1HashThatNeverMatchesAnyPassword = make([]byte, crypto.MasterKeyLength+v1SaltLength)
 
-func (p *Profile) setPasswordV1(password string) error {
+func (p *Profile) setPasswordV1(password, keyDerivationAlgorithm string) error {
 	salt := make([]byte, v1SaltLength)
 	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
 		return errors.Wrap(err, "error generating salt")
 	}
 
+	var err error
+
 	p.PasswordHashVersion = 1
-	p.PasswordHash = computePasswordHashV1(password, salt)
+	p.PasswordHash, err = computePasswordHashV1(password, salt, keyDerivationAlgorithm)
 
-	return nil
+	return err
 }
 
-func computePasswordHashV1(password string, salt []byte) []byte {
-	key, err := scrypt.Key([]byte(password), salt, v1ScryptN, v1ScryptR, v1ScryptP, v1KeyLength)
+func computePasswordHashV1(password string, salt []byte, keyDerivationAlgorithm string) ([]byte, error) {
+	key, err := crypto.DeriveKeyFromPassword(password, salt, keyDerivationAlgorithm)
 	if err != nil {
-		panic("unexpected scrypt error")
+		return nil, errors.Wrap(err, "error deriving key from password")
 	}
 
 	payload := append(append([]byte(nil), salt...), key...)
 
-	return payload
+	return payload, nil
 }
 
-func isValidPasswordV1(password string, hashedPassword []byte) bool {
-	if len(hashedPassword) != v1SaltLength+v1KeyLength {
+func isValidPasswordV1(password string, hashedPassword []byte, keyDerivationAlgorithm string) bool {
+	if len(hashedPassword) != v1SaltLength+crypto.MasterKeyLength {
 		return false
 	}
 
 	salt := hashedPassword[0:v1SaltLength]
 
-	h := computePasswordHashV1(password, salt)
+	h, err := computePasswordHashV1(password, salt, keyDerivationAlgorithm)
+	if err != nil {
+		panic(err)
```

Comment:
This should propagate the error instead of calling `panic`king.

Previously, computePasswordHashV1 could panic if `scrypt.Key` returned an error. `scrypt.Key` only returns an error when _"... the parameters do not satisfy the limits"_ https://pkg.go.dev/golang.org/x/crypto/scrypt#Key
The parameters specified to `scrypt.Key` were constants within the valid range, thus an error was expected to never be returned, and the panic was an "impossibility".

In this implementation, it is possible for `crypto.DeriveKeyFromPassword` to return an error for other conditions, such an invalid key derivation algorithm. This means, that the `panic` here is no longer an "impossibility", and instead the error should be handled or propagated.

This is addressed in #3907